File: src/main/java/dp/combination/combination.cpp
#include <iostream>
using namespace std;

class Solution {

    void helper(int n, int idx, int k, vector<vector<int>>& res, vector<int>& temp) {
        //base
        if(k==0) {
            res.push_back(temp);
            return;
        }

        for(int i=idx; i<=n; i++) {
            temp.push_back(i);
            helper(n, i+1, k-1, res, temp);
            temp.pop_back();
        }
    }

public:
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>> res;
        vector<int> temp;
        helper(n, 1, k, res, temp);
        return res;
    }
};

int main()
{
    Solution s;
    vector<vector<int>> res =  s.combine(4, 3);

    for(vector<int> row: res) {
        cout<<"(";
        for(int s: row) {
            cout<<s<<", ";
        }
        cout<<")"<<endl;
    }

    return 0;
}

File: src/main/java/dp/combination/combination.py
def fn(n, idx, k, res, temp):
    if k == 0:
        res.append(temp.copy())
        return

    for i in range(idx, n + 1):
        temp.append(i)
        fn(n, i + 1, k - 1, res, temp)
        temp.pop()

fn(4, 1, 3, [], [])

#https://www.recursionvisualizer.com/

# def fn(n, idx, k, res, temp):
#     if k == len(temp):
#         res.append(temp)
#         return
#
#     if idx > n:
#         return
#
#     temp.append(idx)
#     fn(n, idx + 1, k, res, temp)
#     temp.pop()
#
#     fn(n, idx + 1, k, res, temp)
# fn(4, 1, 3, [], [])